content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
+how do we efficiently track and clean up expired fields inside hashes that may contain thousands (or millions) of entries,
+when only a small subset are volatile — and an even smaller fraction are expired at any given time?
+The challenge is balancing three conflicting goals:
+
+* **Keep memory usage small.** Hashes can grow to millions of fields, so metadata overhead must be minimal.
+
+* **Maintain fast lookups and updates.** Most hash commands run in `O(1)` time, and expiration tracking can’t change that.
+
+* **Reclaim memory efficiently.** The active expiration job is time-bounded, so we need to minimize wasted CPU cycles spent scanning unexpired fields.
+
+## Alternatives Considered
+
+We explored several approaches:
+
+### 1. Secondary hashtable per field expiration
+
+This seemed simple: build a secondary hashtable in each Hash object which maps only volatile fields.
+
+![illustration](hfe-alternative-hashtable.png)
+
+This is actually the way volatile generic keys are being tracked. Per each database, Valkey manage a secondary map mapping volatile keys.
+During active expiration process the existing job scan the secondary map and each key found to pass it's assigned TTL is expired and it's memory is reclaimed.
+The problem with this design option is the potential inefficiency introduced while scanning many items which should not expire.
+
+### 2. Radix Tree-based index
+
+Using a radix tree to hold field names plus expirations provides sorted access for free.
+
+![illustration](hfe-alternative-radixtree.png)
+
+This is also not a new concept in Valkey. In fact this exact structure is being used to manage client connections blocked on key operations
+like [`BLPOP`](/commands/blpop/) and [`XREAD`](/commands/xread/).
+
+But the memory overhead per node was high. During experiments we measured more than 54 bytes overhead per each hash field when using this type of index.
+
+### 3. Global sorted structure
+
+We wanted to have the ability to efficiently scan over fields which are already expired. A good way to achieve this is by using a sorted index.
+Using a radix tree was possible, but the memory overhead was high. Instead we could use a more lightweight data structure like a skip list, which memory consumption is more bounded
+and is not governed by the data atrophy.
+
+![illustration](hfe-alternative-skiplist.png)
+
+However, This would give `O(log N)` access to the index which did not work well with our target to keep hash operations constant time complexity.
+
+## Coarse Buckets with Adaptive Encoding
+
+Instead of tracking every field’s expiration individually, we designed a coarse bucket system.
+Each field’s timestamp is mapped into a time bucket, represented by a shared “end timestamp.”
+
+![illustration](hfe-coarse-buckets.png)
+
+This solution introduces a semi-sorted data structure which we named 'vset' (stands for "volatile set").
+The vset manages buckets in different time window resolutions and alternating encodings.
```

Comment:
```suggestion
The vset manages buckets in different time window resolutions and adaptive encodings.
```
Alternating implies it switches back and forth, and I guess it's more consistent with the header.

Reply:
Alternating is like A,B,A,B,A,B. Which I don't think is true.

Reply:
mmm. we switch from single pointer to vector and the other way around. we also switch from hashtable to single... so there are back mutations. but I will use adaptive as the header states.